tools/thanosconvert/thanosconvert.go

```diff
@@ -0,0 +1,188 @@
+	return PerUserResults{
+		FailedBlocks:    []string{},
+		ConvertedBlocks: []string{},
+		UnchangedBlocks: []string{},
+	}
+}
+
+func (r *PerUserResults) AddFailed(blockID string) {
+	r.FailedBlocks = append(r.FailedBlocks, blockID)
+}
+func (r *PerUserResults) AddConverted(blockID string) {
+	r.ConvertedBlocks = append(r.ConvertedBlocks, blockID)
+}
+
+func (r *PerUserResults) AddUnchanged(blockID string) {
+	r.UnchangedBlocks = append(r.UnchangedBlocks, blockID)
+}
+
+type Results map[string]PerUserResults
+
+// NewThanosBlockConverter creates a ThanosBlockConverter
+func NewThanosBlockConverter(ctx context.Context, cfg bucket.Config, dryRun bool, logger log.Logger) (*ThanosBlockConverter, error) {
+	bkt, err := bucket.NewClient(ctx, cfg, "thanosconvert", logger, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ThanosBlockConverter{
+		bkt:    bkt,
+		logger: logger,
+		dryRun: dryRun,
+	}, err
+}
+
+// Run iterates over all blocks from all users in a bucket
+func (c ThanosBlockConverter) Run(ctx context.Context) (Results, error) {
+	results := make(Results)
+
+	scanner := cortex_tsdb.NewUsersScanner(c.bkt, cortex_tsdb.AllUsers, c.logger)
+	users, _, err := scanner.ScanUsers(ctx)
+	if err != nil {
+		return results, errors.Wrap(err, "error while scanning users")
+	}
+	level.Info(c.logger).Log("msg", "Scanned tenants")
+
+	for _, u := range users {
+		r, err := c.convertUser(ctx, u)
+		results[u] = r
+		if err != nil {
+			return results, errors.Wrap(err, fmt.Sprintf("error converting user %s", u))
+		}
+	}
+	return results, nil
+}
+
+func (c ThanosBlockConverter) convertUser(ctx context.Context, user string) (PerUserResults, error) {
+	results := PerUserResults{}
+
+	userBucketClient := bucket.NewUserBucketClient(user, c.bkt)
+	err := userBucketClient.Iter(ctx, "", func(o string) error {
+		// get block ID from path
+		o = strings.TrimSuffix(o, "/")
+		blockULID, err := ulid.Parse(o)
+		if err != nil {
+			// this is not a block, skip it
+			return nil
+		}
+		blockID := blockULID.String()
+
+		level.Debug(c.logger).Log("msg", "Processing block", "block", blockID)
+
+		// retrieve meta.json
+
+		metaKey := fmt.Sprintf("%s/%s/meta.json", user, blockID)
+		r, err := c.bkt.Get(ctx, metaKey)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "Get meta.json", "block", blockID, "meta_key", metaKey, "err", err.Error())
+			results.AddFailed(blockID)
+			return nil
+		}
+
+		meta, err := metadata.Read(r)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "decode meta.json into metadata.Meta", "block", blockID, "meta_key", metaKey, "err", err.Error())
+			results.AddFailed(blockID)
+			return nil
+		}
+
+		// convert and upload if appropriate
+
+		newMeta, changesRequired := convertMetadata(*meta, user)
+
+		if len(changesRequired) > 0 {
+			if c.dryRun {
+				level.Debug(c.logger).Log("msg", "Block requires changes, not uploading due to dry run", "block", blockID, "changes_required", strings.Join(changesRequired, ","))
```

Comment:
Do we need to repeat "not uploading due to dry run" message for each block? Perhaps just short "(dry-run)" would do.